examples/QIR/Simulation/TraceSimulator/README.md

```diff
@@ -0,0 +1,285 @@
+
+```cpp
+static Result zero = reinterpret_cast<Result>(0);
+static Result one = reinterpret_cast<Result>(1);
+```
+
+In order to restrict the trace simulator to straight-line pieces of code, measurement result comparisons are disabled by throwing an error on `AreEqualResults` calls of the interface:
+
+```cpp
+bool TraceSimulator::AreEqualResults(Result r1, Result r2)
+{
+    // Don't implement measurement-based branching for the trace simulator.
+    throw std::logic_error("operation_not_supported");
+}
+```
+
+Such a straight-line code restriction is not a requirement for trace simulators, but it works well with the type of hardware currently available (when using the trace simulator in that context).
+For a much more complete trace simulator that is also used for resource estimation, see the [QIR Runtime trace simulator](https://github.com/microsoft/qsharp-runtime/tree/main/src/Qir/Runtime/lib/Tracer).
+
+---
+
+`TraceSimulation.cpp`
+
+Most of the instruction set required by the `IQuantumGateSet` interface consists of single-qubit gates and multi-controlled single-qubit gates.
+Thus it makes sense to reuse two private functions to print out these gates:
+
+```cpp
+void TraceSimulator::ApplyGate(Gate gate, Qubit target)
+{
+    std::cout << "Applying gate \"" << gate << "\" on qubit "
+              << this->qbm->GetQubitName(target) << std::endl;
+}
+
+void TraceSimulator::ApplyControlledGate(Gate gate, long numControls, Qubit controls[], Qubit target)
+{
+    std::cout << "Applying gate \"" << gate << "\" on target qubit "
+              << this->qbm->GetQubitName(target) << " and controlled on qubits ";
+    for (int i = 0; i < numControls; i++)
+        std::cout << this->qbm->GetQubitName(controls[i]) << " ";
+    std::cout << std::endl;
+}
+```
+
+The text that is printed can be adjusted to suit the specific needs of the application.
+The individual gates from the instruction set then call the above functions by providing the correct gate name:
+
+```cpp
+void TraceSimulator::X(Qubit q)
+{
+    ApplyGate("X", q);
+}
+
+void TraceSimulator::ControlledX(long numControls, Qubit controls[], Qubit target)
+{
+    ApplyControlledGate("X", numControls, controls, target);
+}
+```
+
+The multi-qubit gates `Exp` and `ControlledExp` can be handled similarly, just with additional printing support for multiple target qubits and different Pauli bases, and similarly for the `Measure` instruction.
+
+An example output might look as follows:
+
+```output
+Applying gate "H" on qubit 0
+Applying gate "X" on target qubit 1 and controlled on qubits 0
+Applying gate "R(1.570796)_Z" on qubit 0
+Applying gate "Exp(0.100000, X Y)" on qubits 0 1
+Measuring qubits:
+    0 in base Z
+    1 in base X
+```
+
+## Compiling the simulator
+
+The simulator samples require a working [Clang](https://clang.llvm.org/) installation to compile.
+Refer to the [Optimization example](../Optimization#installing-clang) for instructions on setting up Clang and LLVM.
```

Comment:
@dime10 This line seems to be the source of the broken link. Looks like it's missing an extra `..` and should be:
```suggestion
Refer to the [Optimization example](../../Optimization#installing-clang) for instructions on setting up Clang and LLVM.
```